계산기 [STEP 1] 레옹아범

fatherLeon opened the pull request:
@Judy-999 
안녕하세요 쥬디 계산기 첫 PR보냅니다 프로젝트 기간동안 잘 부탁드립니다🙇
STEP1정확한 목표가 어디까지인지 몰라서 Queue까지 구현하면 끝나는 내용? 인것 같아 PR보냅니다.....
CalculatorItemQueue의 몇몇 메소드 주석은 STEP2에서 구현하면서 지우도록 하겠습니다

## 고민한점
### 각각의 모델을 클래스로 만들지?, 구조체로 만들지?
> [Apple Developer Documentation](https://developer.apple.com/documentation/swift/choosing-between-structures-and-classes)
* 해당 문서를 통해 기본적인 default는 구조체로 잡고 시작했습니다.
* 다만 Node의 경우 Class를 만들었는데 LinkedList특성상 해당 Node의 prev와 next가 Node를 복사하기보다는 참조한다는 개념으로 접근하는게 맞는것 같아 Node만 클래스로 작성하였습니다

### Queue만을 쓸지 Queue내에 새로운 자료구조를 추가할지
* 우선 Queue의 구현에 대해 고민해봤는데 어떤식으로 구현을 하든 enqueue와 dequeue 둘 중 하나는 O(n)의 시간복잡도를 가지게 되는걸 확인했습니다.
* 다만, Queue안에 LinkedList를 넣을 경우 데이터를 추가할 떄도, 뺄 때도 O(1)이라는 시간복잡도를 가지기 때문에 Queue내에 LinkedList를 추가하기로 했습니다.

### 어떤 형태로 모델을 설계해야하는지?
* UML을 작성하다보니 CalculatorItemQueue에 enqueue, dequeue 그 외의 queue관련 메소드를 작성하려고 보니 추후에 계산을 하는 함수 값을 지우는 함수등을 작성하다보면 CalculatorItemQueue가 너무 커질것 같아서 Queue라는 객체를 따로 만들어 그 안에 Queue관련 메소드와 프로퍼티를 넣고 CalculatorItemQueue에는 추후 계산기 관련 메소드 및 Queue타입의 프로퍼티가 들어갈 수 있도록 만들었습니다

### LinkedList의 모든 내용을 삭제하고 싶을 때
```swift
struct Queue {
    mutating func removeAll() {
        self.head = nil
        self.tail = nil
    }
}
```
* 계산기의 AC를 누를 경우 다 삭제한다는 내용이 있어서 Queue의 LinkedList를 삭제할려고 하니 각각의 연결된 노드가 서로 prev와 next로 연결되어 있어 강한참조 문제가 있을것 같아서 아래와 같이 Node의 프로퍼티에 weak를 추가하였습니다.

```swift
final class Node {
    weak var prev: Node?
    weak var next: Node?
}
```

## 조언을 얻고싶은점 && 해결하지 못한 점
### 테스트케이스 객체 생성관련
#### QueueTest
```swift
private func enqueueNodes(_ nodes: Node...) {
    for node in nodes {
        sut?.enqueue(node)
    }
}
```

* 해당 부분을 통해서 Node를 enqueue해주고 있습니다.

```swift
func test_Queue에_node를_하나씩_넣는다면_tail에_들어간다() {
    let firstNode = Node(value: "1")
    let secondNode = Node(value: "2")
        
    enqueueNodes(firstNode, secondNode)
        
    XCTAssertEqual(secondNode.value, sut!.tail!.value)
}
```

* 그리고 이부분에서 Node인스턴스를 만들어주는데 이럴경우 테스트 더블의 종류 중 Dummy나 Stub을 만들어서 추가해야할까요???

fatherLeon committed:
init: 모델 파일 생성

Protocol - CalculateItem 생성
Model - Queue, Node, CalculatorItemQueue 생성

init

fatherLeon committed:
init: step1 클래스 다이어그램 생성

Node, Queue, CalculatorItemQueue, CalculateItem관련 클래스 다이어그램 생성

init

fatherLeon committed:
feat: Node 구현

fatherLeon committed:
feat: Queue 구현

Node를 통해 연결리스트로 만든 Queue 구현

feat

fatherLeon committed:
chore: 주석 변경

상단의 파일 설명 주석 변경

chore

fatherLeon committed:
feat: CalculatorItemQueue 구현

removeAllItems 구현

feat

fatherLeon committed:
docs: UML수정

클래스 다이어그램 프로퍼티 및 메소드 추가

docs

fatherLeon committed:
chore: 함수명 수정

Judy-999 commented on Calculator/Calculator.xcodeproj/project.pbxproj:
```diff
@@ -37,10 +99,75 @@
+			isa = PBXGroup;
+			children = (
+				C713D9452570E5EB001C3AFC /* ViewController.swift */,
+				C713D9432570E5EB001C3AFC /* SceneDelegate.swift */,
+				C713D9412570E5EB001C3AFC /* AppDelegate.swift */,
```
그룹으로 파일을 분류하신 점 좋네요 👍

다만 `Controller` 그룹으로 되어 있는데 `SceneDelegate`와 `AppDelegate`가 Controller의 역할을 하는걸까요 🤔

fatherLeon replied:
## SceneDelegate
* 화면에 무엇을 보여줄지 관리하는 역할
* 만약 스토리보드가 없으면 해당 클래스 내에서 첫 뷰 컨트롤러를 설정해준다

## AppDelegate
* 앱 전체의 생명주기를 관리함
* 단순히 뷰가 아닌 scene(뷰가 쌓이는 뷰 계층의 컨테이너 역할)을 부르거나 버리는등의 상황에서 사용

## 정리
* 위의 역할로 봤을 때 단순히 모델과 뷰를 이어주는 순수한 컨트롤러의 역할은 아니지만 앱의 하위단계에서 뷰를 위해서 여러 부분을 담당한다는 점을 봤을 때 어느정도 컨트롤러의 역할도 해준다고 생각합니다

Judy-999 replied:
파일 분류는 개인의 취향이지만 저는 App과 긴밀한 과련이 있는 파일이라 `Application`이라고 따로 분리하는 편입니다!

Judy-999 commented on Calculator/Calculator/Model/Protocol/CalculateItem.swift:
```diff
@@ -0,0 +1,6 @@
+//  CalculateItem.swift
+//  Created by 강민수 on 2023/01/24.
+
+protocol CalculateItem {
+    
+}
```
`CalculateItem`은 어떤 의미와 역할을 하는 프로토콜일까요?
어떤 타입들이 이 프로토콜을 채택해야 할지 고민해보시면 좋겠습니다!

fatherLeon replied:
#### CalculatorItemQueue에서 다루는 요소는 CalculateItem 프로토콜을 준수합니다.
* 해당 글을 다시 읽어보니 계산기의 클릭에 따라서 알맞은 연산자 혹은 숫자를 나타낼 수 있도록 역할을 하는 프로토콜 같습니다
* String, Double, Int와 같이 Node의 value값에 들어가는 타입이 채택해야하는 프로토콜 같습니다

Judy-999 commented on Calculator/Calculator/Model/Node.swift:
```diff
@@ -0,0 +1,14 @@
+//  Node.swift
+//  Created by 레옹아범 on 2023/01/24.
+
+final class Node {
```
상속하지 않는 클래스에 `final` 아주 좋네요 👍

Judy-999 commented on Calculator/Calculator/Model/Node.swift:
```diff
@@ -0,0 +1,14 @@
+//  Created by 레옹아범 on 2023/01/24.
+
+final class Node {
+    weak var prev: Node?
+    weak var next: Node?
```
`weak` 키워드로 순환 참조까지 생각해보셨네요 👏

혹시 `weak`가 언제 해제되는지 알고 계실까요? 순환 참조 문제는 없지만 서로를 약하게 참조하고 있는 것도 문제가 될 수 있을 것 같은데 고민해보시고 알려주세요!

Judy-999 replied:
다음 노드만 알고 있어도 연결 리스트의 역할을 할 수 있을 텐데
이전 노드와 다음 노드를 참조하고 있는 이중 연결 리스트로 구현하신 이유가 있을까요?

fatherLeon replied:
### weak의 문제점
* 현재 head와 tail이 Node에 대한 인스턴스를 가지고 있기 때문에 단순히 추가할때는 문제가 없지만, head나 tail의 Node 참조를 제거하면 head tail 이외의 Node는 약한참조이기 때문에 메모리에서 deallocated된다는 문제점이 생길 것 같습니다

### 이중연결리스트
* 예를들어 tail값을 제거 한다거나 중간의 Node를 제거하는 상황이 생긴다면 단일연결리스트보다 이중연결리스트가 구현이나 실행속도나 더 빠를것 같아 이중연결리스트로 구현을 했습니다.
* 그렇지만 Queue 특성상 값을 추가 한다면 tail에 추가를하고 값을 제거한다면 head의 값만 제거하면 되기 때문에 단일연결리스트로 새롭게 구현하였습니다!!

Judy-999 commented on Calculator/Calculator/Model/Queue.swift:
```diff
@@ -0,0 +1,48 @@
+
+struct Queue {
+    var head: Node?
+    var tail: Node?
```
`Queue`라는 타입의 역할은 무엇인가요?

요구사항에 보면 `사칙연산 계산을 위한 큐(Queue) 타입(CalculatorItemQueue)`이라고 되어 있는데, 큐를 따로 구현하는 것이 아니라 `CalculatorItemQueue` 자체가 큐여야 할 것 같네요!

Judy-999 replied:
그리고 `head`와 `tail`을 가진 것으로 보아 큐가 아니라 연결 리스트에 가깝지 않나 싶네요 🙂

fatherLeon replied:
* 말씀을 들어보니 큐 보다는 연결리스트인것 같고 CalculatorItemQueue에 큐를 구현하는게 맞는것 같습니다
* Queue파일을 -> LinkedList로 바꾸고 기존 이중연결리스트에서 단일연결리스트로 새로 구현하였습니다

Judy-999 commented on Calculator/Calculator/Model/Queue.swift:
```diff
@@ -0,0 +1,48 @@
+        node.prev = self.tail
+        self.tail?.next = node
+        self.tail = node
+        self.tail?.next = self.head
```
`tail`이 `head`를 가리키고 있는 원형 연결 리스트인 걸까요?
원형 연결 리스트라면 그렇게 구현하신 이유는 무엇일까요~?

fatherLeon replied:
* 원형 연결리스트로 구현시 어떠한 Node로 접근을 하던지 모든 Node를 조회할 수 있다는 부분때문에 원형 연결리스트로 구현했습니다
* 다만, 큐라는 자료구조 특성상 head의 값을 제거하거나 확인하기 때문에 원형 연결리스트가 아닌 단일연결리스트로 다시 구현하였습니다!!

Judy-999 commented on Calculator/Calculator/Model/Queue.swift:
```diff
@@ -0,0 +1,48 @@
+            self.tail = node
+            
+            return
+        }
```
외부에서 `Node` 타입을 직접 생성해서 넘겨주는 것 보다는 필요한 값만 넘겨주면 `Node`로 만들어 연결하는 방식은 어떤가요?👀

fatherLeon replied:
#### LInkedLIst
```swift
mutating func append(_ element: Element) {
    if isEmpty {
        self.head = Node(value: element)
        self.tail = self.head
        return
    }
        
    let node = Node(value: element)
    self.tail?.next = node
    self.tail = node
}
```

* 기존 Node를 받아 tail에 추가하는 방법에서 CalculateItem 프로토콜을 채택한 타입을 함수 인자로 받아서 tail에 추가하는 방법으로 개선하였습니다

Judy-999 commented on Calculator/Calculator/Model/Queue.swift:
```diff
@@ -0,0 +1,48 @@
+        dequeueNode?.next = nil
+        
+        return dequeueNode
+    }
```
외부에서 원하는건 `value`일 텐데 역시 `Node`를 반환하는 것 보다는 값을 반환하는게 좋을 것 같습니다.

fatherLeon replied:
#### LinkedList
```swift
mutating func removeFirst() -> Element? {
    if isEmpty {
        return nil
    }
        
    let value = self.head?.value
    self.head = self.head?.next
        
    return value
}
```
* 이 부분도 append와 유사하게 CalculateItem 프로토콜을 채택한 타입을 반환하도록 만들었습니다.

Judy-999 commented on Calculator/Calculator/Model/Queue.swift:
```diff
@@ -0,0 +1,48 @@
+//  Queue.swift
+//  Created by 레옹아범 on 2023/01/24.
+
+struct Queue {
```
모든 메서드에 `mutating`을 붙여주면서도 `struct`로 구현하신 이유가 있으실까요?

레옹아범은 어떤 기준으로 Struct와 Class를 선택하는지 궁금하네요!

fatherLeon replied:
> https://developer.apple.com/documentation/swift/choosing-between-structures-and-classes

* 위 내용을 토대로 기본적으로 구조체로 구현을 하였습니다
* Node의 경우 연결이 값보다는 주소를 통해서 연결이 되어야 하기 때문에 Class로 구현을 하였다면
* 지금 LinkedList의 경우 objective-c관련 내용을 사용하는것도, Node처럼 참조가 필요한것도, 다른 클래스를 상속하는것도 아니어서 구조체로 구현하였습니다

Judy-999 commented on Calculator/Calculator/Model/Node.swift:
```diff
@@ -0,0 +1,14 @@
+final class Node {
+    weak var prev: Node?
+    weak var next: Node?
+    var value: String
```
값(value)으로 `String` 타입만 받을 수 있는데 다양한 타입을 수용할 수 있도록 범용성을 높이는 방법도 고민해보시면 좋겠습니다~

fatherLeon replied:
#### Node
```swift
final class Node<Element: CalculateItem> {
}
```

#### LinkedList
```swift
struct LinkedList<Element: CalculateItem> {
}
```

#### CalculatorItemQueue
```swift
struct CalculatorItemQueue<Element: CalculateItem> {
}
```

* 세가지 타입 모두 Element로 CalculateItem을 채택한 타입이라면 모든 타입을 받을 수 있게 만들어줬습니다

```swift
extension String: CalculateItem {
}
```
* 테스트시에는 String에 CalculateItem을 채택하여 테스트하였습니다

Judy-999 replied:
제네릭과 `CalculateItem` 프로토콜을 잘 활용하셨네요 👍

Judy-999 commented on Calculator/Calculator/Model/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,43 @@
+
+struct CalculatorItemQueue: CalculateItem {
+    var items = Queue()
+    var item: String = ""
```
큐인 `CalculatorItemQueue`가 큐 타입을 갖고 있는건 어색해보여요.
요구사항에서 `CalculatorItemQueue의 내부에서 사용하는 List 형식의 자료구조..`라고 되어 있으니 List 형식의 자료구조를 이용해 큐를 구현하는게 맞을 것 같네요!

Judy-999 replied:
역시 큐인 `CalculatorItemQueue`가 아이템이라는 의미인 `CalculateItem`을 채택하는게 맞는지 고민해보시면 좋겠습니다~

fatherLeon replied:
#### CalculatorItemQueue
```swift
struct CalculatorItemQueue<Element: CalculateItem> {
}
```
* Int, String, Double과 같은 타입이 CalculateItem을 채택하는 방법이 맞는것 같아 수정하였습니다

Judy-999 commented on Calculator/Calculator/Model/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,43 @@
+//        2+3-4+5를 입력하고 CE를 누르면 2+3-4+의 상태가 됩니다.
+//        2+3-4+ 를 입력하고 CE를 누르면 2+3-4+의 상태가 됩니다.
+//        2+3-4+5= 을 입력하고 CE를 누르면 0의 상태가 됩니다.
+    }
```
계산기를 위한 큐라고 생각하셔서 메서드의 형태들을 잡아두신 것 같은데 큐는 `enqueue`, `dequeue`기능을 제공하는 자료구조입니다.

큐 자체는 큐의 역할만 하고 외부에서 큐의 특성을 이용해 계산을 하는 방식으로 구현되어야 하지 않을까 생각이 드네요 🤔
고민해보세요💡

fatherLeon replied:
#### CalculatorItemQueue
```swift
struct CalculatorItemQueue<Element: CalculateItem> {
    private var items = LinkedList<Element>()
    
    var isEmpty: Bool {
        self.items.isEmpty
    }
    var peek: Element? {
        self.items.head?.value
    }
    var back: Element? {
        self.items.tail?.value
    }
    
    mutating func enqueue(_ element: Element) {
        self.items.append(element)
    }
    
    mutating func dequeue() -> Element? {
        return self.items.removeFirst()
    }
    
    mutating func removeAll() {
        self.items.removeAll()
    }
}
```
* 위와 같이 큐에서 필요하고 사용되는 기능만을 넣어 구현하였습니다

Judy-999 commented on Calculator/NodeTest/NodeTest.swift:
```diff
@@ -0,0 +1,34 @@
+        sut?.value = "+ " + (sut?.value)!
+        
+        XCTAssertEqual(value, sut?.value)
+    }
```
테스트는 기능(메서드)을 확인해야 한다고 생각합니다.

Node는 메서드를 가지고 있지 않는데 해당 테스트들을 `init`이 잘 동작하는지, 프로퍼티가 `var`인지 확인하는 수준으로 보여 불필요한 테스트는 아닌가 싶네요 👀

Judy-999 commented on Calculator/CalculatorItemQueueTest/CalculatorItemQueueTest.swift:
```diff
@@ -0,0 +1,47 @@
+
+    override func tearDownWithError() throws {
+        sut = nil
+    }
```
`override`로 재정의할 때는 잊지 말아야 할 단계가 있습니다~! 확인해주세요!

Judy-999 commented on Calculator/Calculator/Model/Protocol/CalculateItem.swift:
```diff
@@ -0,0 +1,6 @@
+//  CalculateItem.swift
+//  Created by 강민수 on 2023/01/24.
```
사소하지만 다른 파일은 레옹아범으로 되어 있는데 수정이 안된 것 같네요!

Judy-999 commented on Calculator/QueueTest/QueueTest.swift:
```diff
@@ -0,0 +1,161 @@
+        }
+    }
+    
+    func test_Queue가_처음_생성되면_head와_tail이_nil이다() {
```
테스트 메서드의 이름을 한글로 하신 이유는 무엇일까요~?

Judy-999 replied:
Unit Test는 given-when-then으로 분리해서 코드를 작성하곤 합니다. 이를 주석으로 구분하는 것도 좋습니다!

fatherLeon replied:
### 메서드 이름을 한글로 한 이유
* 영어로 정확하게 적을 자신도 없고... 한글이 어떤 상황에서 어떤 동작을하고 무엇을 확인하는지에 대해 조금 더 명확하게 적을 수 있기 때문에 한글로 작성하였습니다

### given/when/then
* 주석을 통해서 given/when/then(준비-실행-검증)단계를 나눠줬습니다

Judy-999 commented on Calculator/QueueTest/QueueTest.swift:
```diff
@@ -0,0 +1,161 @@
+        XCTAssertEqual(sut?.head == nil, sut?.tail == nil)
+    }
+    
+    func test_Queue가_비어있을때_isEmpty프로퍼티는_true이다() {
+        XCTAssertTrue(sut!.isEmpty)
+    }
```
`비어있을 때~` 라는 건 명확한 상황이 아닌 것 같아요! 처음 생성했을 때를 의도하신걸까요?

Judy-999 replied:
Node를 모두 dequeue한 후에 isEmpty가 true인지도 확인해보시면 좋겠네요 🙂

fatherLeon replied:
* ```비어있을 때~```와 같은 테스트케이스는 첫 생성을 했을 때를 나타내기 위한 의도입니다
* 다만 불분명한것 같아
```swift
func test_초기_인스턴스_생성시_isEmpty는_true이다() {
    // then
    XCTAssertTrue(sut.isEmpty)
}
```
* 초기 인스턴스 생성시라는 말을 넣어줬으며 그 외에는 굳이 이런말이 필요 없이 각각의 테스트케이스 마다 항상 첫 생성했을 때를 의미하므로 적지 않았습니다

fatherLeon replied:
* 모두 dequeue한 이후 head와 tail값이 nil인지, isEmpty가 true인지도 테스트케이스에 추가하였습니다
* 해당 테스트케이스를 통해 tail에 값이 남아있는 것을 확인해 수정하였습니다!!!

Judy-999 commented on Calculator/QueueTest/QueueTest.swift:
```diff
@@ -0,0 +1,161 @@
+        XCTAssertTrue(sut!.isEmpty)
+    }
+    
+    func test_Queue가_비어있지않으면_isEmpty프로퍼티는_false이다() {
+        sut?.head = Node(value: "b")
+        
+        XCTAssertFalse(sut!.isEmpty)
```
역시 `비어있지않으면`은 명확한 상황이 아닌 것 같습니다~

fatherLeon replied:
```swift
func test_a의_값을_append하면_isEmpty는_false이다() { 
}
```
* 이런 식으로 준비-실행-검증 단계를 적었습니다

Judy-999 commented on Calculator/QueueTest/QueueTest.swift:
```diff
@@ -0,0 +1,161 @@
+import XCTest
+@testable import Calculator
+
+final class QueueTest: XCTestCase {
+    
+    var sut: Queue?
```
CalculatorItemQueue 테스트에서는 아래와 같이 `!`를 사용하셨는데 이곳에선 `?`를 사용하신 이유가 있을까요?
```swift
var sut: CalculatorItemQueue!
```

fatherLeon replied:
오타입니다 ㅠ
테스트케이스에서는 !와 같이 강제언래핑을 하는것이 코드 작성에도 유리하고 코드도 명확해서 !(강제언래핑)으로 변경하였습니다

Judy-999 commented on Calculator/QueueTest/QueueTest.swift:
```diff
@@ -0,0 +1,161 @@
+        XCTAssertEqual(sut!.head!.value, sut!.tail!.value)
+    }
+    
+    func test_Queue에_node를_하나씩_넣는다면_tail에_들어간다() {
+        let firstNode = Node(value: "1")
+        let secondNode = Node(value: "2")
+        
+        enqueueNodes(firstNode, secondNode)
+        
+        XCTAssertEqual(secondNode.value, sut!.tail!.value)
+    }
```
저라면 명확하게 표현하기 위해 다음과 같이 명명했을 것 같아요 🙂
`하나씩_넣는다면` --> 1과 2를 넣는다면
`tail에_들어간다` --> tail의 값은 2가 된다

다른 테스트의 이름들도 모호하진 않은지 확인해주세요!

Judy-999 replied:
또한 어떤 메서드를 테스트하는 것인지도 드러나면 좋겠네요!

fatherLeon replied:
```swift
func test_a의_값을_append한다면_tail값은_a이다() { }
func test_a과_b_두개의_값을_append한다면_head는_a_tail은_b다() { }
```
* 이렇게 최대한 구체적으로 적었습니다
* 그리고 각각의 프로퍼티와 메소드 이름까지 작성하여 어떤것을 실행하고 검증하는지 명확히 할려고 했습니다

Judy-999 commented on Calculator/QueueTest/QueueTest.swift:
```diff
@@ -0,0 +1,161 @@
+        XCTAssertEqual(secondNode.value, sut!.tail!.value)
+    }
+    
+    func test_Queue가_2개일_경우_head의_next는_tail을_가리킨다() {
+        let firstNode = Node(value: "1")
+        let secondNode = Node(value: "2")
+        
+        enqueueNodes(firstNode, secondNode)
+        
+        let headNextValue = sut!.head!.next!.value
+        let tailValue = sut!.tail!.value
+        
+        XCTAssertEqual(headNextValue, tailValue)
+    }
+    
+    func test_Queue가_2개일_경우_tail의_prev는_head를_가리킨다() {
+        let firstNode = Node(value: "1")
+        let secondNode = Node(value: "2")
+        
+        enqueueNodes(firstNode, secondNode)
+        
+        let tailPrevValue = sut!.tail!.prev!.value
+        let headValue = sut!.head!.value
+        
+        XCTAssertEqual(tailPrevValue, headValue)
```
`Queue가 몇 개 일 경우~`라는 것은 현재 Queue가 가지고 있는 Node가 몇 개인지를 의미하는 것 같은데 큐가 여러 개 있는 걸로 오해할 것 같습니다 😅

Judy-999 replied:
`prev`와 `head`가 무엇을 가리키는 것이 중점이 아닌 메서드(enqueue, dequeue 등)가 잘 수행되는지 여러 시나리오로 테스크가 수행되어야 할 것 같아요!

fatherLeon replied:
```swift
func test_a와_b를_enqueue한다면_dequeue의_값은_a이다() { }
```
* 위와 같이 최대한 명확하게? 쓸려고 노력했습니다

```swift
func test_a와_b를_enqueue하고_dequeue를_한다면_peek의_값은_b이다() { }
```
* prev와 head가 무엇을 가리키는지가 아닌 peek와 back으로 해당 값을 중점적으로 확인했습니다!!

Judy-999 reviewed (changes requested):
안녕하세요 레옹아범 로직의 세계 계산기 프로젝트로 오신걸 환영합니다  🤗

UML, Queue, Unit Test 등 낯선 개념이 많으셨을 텐데 수고하셨습니다 👏👏 
어려우시더라도 차근 차근 이해해보면서 잘 진행해봅시다 🙌

그리고 첫 리뷰이니 당부의 말씀을 드릴게요!

제가 드린 코멘트에 댓글로 어떻게 반영하셨는지 또는 의견을 달아주시면 제가 수정사항을 파악하기 용이합니다. 만약 큰 의미가 없는 코멘트라면 이모지만 달아주셔도 됩니다! 만약 제가 드린 코멘트가 잘 이해가 안가신다면 언제든 DM 주세요~
그리고 PR과 마찬가지로 모든 수정을 마친 후 DM을 주시면 빠르게 확인하겠습니다.

제가 질문드리는 점은 반드시 고쳐야 한다고 둘러서 말하는게 아니라 정말 궁금하거나 의견을 묻는 경우도 있으니 반드시 수정(적용)하거나 잘못된 점이라고 생각하시지 않으면 좋겠네요!
제 의견을 무조건 수용하시기보다 레옹아범의 이유로 결정하는게 더 좋습니다 🙂

저는 레옹아범의 코드를 평가하거나 잘못된 점을 지적하는 사람이 아닙니다!
같이 코드를 고민하는 역할로 더 좋은, 레옹아범의 이유가 있는 코드를 완성하는 데에 도움이 되면 좋겠습니다 😊

---

## 고민한점
> 각각의 모델을 클래스로 만들지?, 구조체로 만들지?

이 문제는 매번 어려우면서도 앞으로도 새로운 타입을 생성할 때마다 고민하셔야 할 문제입니다 🥲 그러니 레옹아범만의 기준을 정해놓고 적용하시면 좋겠습니다!

<br>

> Queue만을 쓸지 Queue내에 새로운 자료구조를 추가할지

Queue만을 사용한다는 것은 어떤 의미일까요? 
알아보신 다른 방법에는 어떤 장단점이 있었는지도 궁금하네요!

더불어 LinkedList에도 여러 방식이 있는데 어떤 이유로 현재 방식을 선택했는지 궁금합니다 🙂
<br>

> 어떤 형태로 모델을 설계해야하는지?

코멘드로도 남겼지만 `CalculatorItemQueue`는 큐의 역할만 하고 계산기 작업은 이후 스텝에서 큐의 특성을 활용해 작업해야 할 것 같아요. 현재는 Queue와 LinkedList의 관계를 먼저 생각해보고 구조를 구성해보시는게 좋을 것 같습니다.
<br>

> LinkedList의 모든 내용을 삭제하고 싶을 때

Node의 프로퍼티 weak에 대한 내용은 코멘드로 남겼습니다!
<br>

## 조언을 얻고싶은점 && 해결하지 못한 점
> 이부분에서 Node인스턴스를 만들어주는데 이럴경우 테스트 더블의 종류 중 Dummy나 Stub을 만들어서 추가해야할까요???

코멘트로도 남겼지만 우선 `enqueue`, `dequeue` 메서드 자체가 Node를 생성해서 넣어주고 반환하는 것 보단 `value`를 사용해야 할 것 같습니다.

사용자가 Queue에서 값을 넣고 뺄 때 궁금한 건 해당 노드가 아니라 노드에 들어있는 값일 것이기 때문이죠.

기회가 되신다면 Test Double에 대해서도 학습해보시면 좋습니다 🙂

---

\+ UML도 PR에 같이 넣어주시면 제가 확인하기 용이합니다~🙌

fatherLeon committed:
chore: 주석 변경

fatherLeon committed:
test: Node 테스트코드 작성

기존 초기값에 새로운 값이 할당가능한지에 대한 테스트

test

fatherLeon committed:
refactor: Node 프로퍼티 및 요소 수정

prev를 통해 이중연결리스트 Node가 아닌
단일 연결리스트 Node로 생성
또한 값에 CalculateItem프로토콜을 채택한 타입을 받을 수 있도록 설정

refactor

fatherLeon committed:
chore: 파일 이름 변경

Queue -> LinkedList
QueueTest -> 삭제

chore

fatherLeon committed:
test isEmpty 및 append 테스트케이스 작성

fatherLeon committed:
test: LinkedList 테스트케이스 추가

fatherLeon committed:
refactor: NodeTest 테스트케이스 이름 명확하게 변경

fatherLeon committed:
test: LinkedList removeLast에 관한 테스트 케이스 작성

fatherLeon committed:
refactor: removeLast -> removeFirst로 이름 변경

fatherLeon committed:
test: removeFirst로 이름변경에 따라 테스트케이스 내용 변경

fatherLeon committed:
feat: LinkedList removeFirst구현

fatherLeon committed:
feat: CalculatorItemQueue enqueue구현

fatherLeon committed:
test: CalculatorItemQueueTest 테스트케이스 추가

enqueue, dequeue테스트케이스 추가

test

fatherLeon committed:
test: CalculatorItemQueue의 removeAll메소드 테스트케이스 작성

fatherLeon committed:
feat: CalculatorItemQueue의 removeAll구현

fatherLeon committed:
refactor: Node타입 수정

Node내의 value가 CalculateItem타입을 가질 수 있게 설정

refactor

fatherLeon committed:
Revert "refactor: Node타입 수정"

This reverts commit 975683184009f78b47f6269032b362923928344f.

fatherLeon committed:
fix: LinkedList removeFirst메소드 수정

tail에 값이 남아있는 문제를 해결
head와 tail의 값이 nil인지에 대한 테스트케이스
isEmpty가 true인지에 대한 테스트 케이스 작성

fix

fatherLeon committed:
refactor: 테스트케이스 이름 수정

fatherLeon committed:
refactor: 테스트케이스 이름 수정

fatherLeon committed:
docs: UML수정

fatherLeon commented:
![](https://raw.githubusercontent.com/fatherLeon/ios-calculator-app/step1/Step1UML.png)

* 코드 수정내용에 따라 UML도 수정하였습니다!!

Judy-999 commented on Calculator/LinkedListTest/LinkedListTest.swift:
```diff
@@ -0,0 +1,144 @@
+import XCTest
+@testable import Calculator
+
+extension String: CalculateItem {
+}
```
테스트용으로만 사용하는게 아니라면 테스트 파일이 아닌 곳에 위치하는게 좋겠습니다!

Judy-999 commented on Calculator/LinkedListTest/LinkedListTest.swift:
```diff
@@ -0,0 +1,144 @@
+    }
+    
+    func test_a와_b_두개의_Node를_가진_LinkedList에서_removeFirst를_한다면_반환값은_a이다() {
+        // given
+        let firstValue = "a"
+        let secondValue = "b"
+        let exception = "a"
+        // when
+        sut.append(firstValue)
+        sut.append(secondValue)
+        let removedValue = sut.removeFirst()!
```
참고로 두 개뿐이긴 하지만 많아진다면 아래와 같은 방식으로 할 수 있습니다~
```suggestion
        // given
        let values = ["a", "b"]
        let exception = "a"
        // when
        values.forEach {
            sut.append($0)
        }
        let removedValue = sut.removeFirst()!
```

Judy-999 commented on Calculator/Calculator/Model/LinkedList.swift:
```diff
@@ -0,0 +1,40 @@
+//  Created by 레옹아범 on 2023/01/24.
+
+struct LinkedList<Element: CalculateItem> {
+    var head: Node<Element>?
+    var tail: Node<Element>?
```
외부에서 수정 가능하도록 노출이 되어 있네요! 적절한 접근 제어를 넣어주세요 🙂

Judy-999 reviewed (approved):
수고하셨어요 👏👏
추가로 남긴 코멘트는 다음 스텝 진행하시면서 반영해주세요! 화이팅 🙌
